addons/purchase_stock/models/res_partner.py

```diff
@@ -17,24 +18,31 @@ class ResPartner(models.Model):
 
     @api.depends('purchase_line_ids')
     def _compute_on_time_rate(self):
-        order_lines = self.env['purchase.order.line'].search([
+        order_line_ids = self.env['purchase.order.line']._search([
             ('partner_id', 'in', self.ids),
             ('date_order', '>', fields.Date.today() - timedelta(365)),
             ('qty_received', '!=', 0),
-            ('order_id.state', 'in', ['done', 'purchase'])
-        ]).filtered(lambda l: l.product_id.sudo().product_tmpl_id.type != 'service')
+            ('order_id.state', 'in', ['done', 'purchase'])])
+
         lines_qty_done = defaultdict(lambda: 0)
-        moves = self.env['stock.move'].search([
-            ('purchase_line_id', 'in', order_lines.ids),
-            ('state', '=', 'done')]).filtered(lambda m: m.date.date() <= m.purchase_line_id.date_planned.date())
-        for move, qty_done in zip(moves, moves.mapped('quantity_done')):
-            lines_qty_done[move.purchase_line_id.id] += qty_done
         partner_dict = {}
-        for line in order_lines:
-            on_time, ordered = partner_dict.get(line.partner_id, (0, 0))
-            ordered += line.product_uom_qty
-            on_time += lines_qty_done[line.id]
-            partner_dict[line.partner_id] = (on_time, ordered)
+
+        for order_lines in split_every(1000, order_line_ids, self.env['purchase.order.line'].browse):
+            order_lines = order_lines.filtered(lambda l: l.product_id.sudo().product_tmpl_id.type != 'service')
+            moves = self.env['stock.move'].search([
```

Comment:
you can use read_group to compute quantity_done (don't need put in cache any stock.move)

Reply:
`quantity_done` is a non stored computed field

Reply:
you can read_group on stock_move_line dom = (move_id.purcharce_line_id, in, ...)